cli/internal/lockfile/pnpm_lockfile.go

```diff
@@ -0,0 +1,234 @@
+	Patched  bool `yaml:"patched,omitempty"`
+	Prepare  bool `yaml:"prepare,omitempty"`
+
+	// only needed for packages that aren't in npm
+	Name    string `yaml:"name,omitempty"`
+	Version string `yaml:"version,omitempty"`
+
+	LibC       []string `yaml:"libc,omitempty"`
+	Deprecated string   `yaml:"deprecated,omitempty"`
+}
+
+// PackageResolution Various resolution strategies for packages
+type PackageResolution struct {
+	Type string `yaml:"type,omitempty"`
+	// For npm or tarball
+	Integrity string `yaml:"integrity,omitempty"`
+
+	// For tarball
+	Tarball string `yaml:"tarball,omitempty"`
+
+	// For local directory
+	Dir string `yaml:"directory,omitempty"`
+
+	// For git repo
+	Repo   string `yaml:"repo,omitempty"`
+	Commit string `yaml:"commit,omitempty"`
+}
+
+// PatchFile represent a patch applied to a package
+type PatchFile struct {
+	Path string `yaml:"path"`
+	Hash string `yaml:"hash"`
+}
+
+func isSupportedVersion(version float32) error {
+	supportedVersions := []float32{5.3, 5.4}
+	for _, supportedVersion := range supportedVersions {
+		if version == supportedVersion {
+			return nil
+		}
+	}
+	return errors.Errorf("Unable to generate pnpm-lock.yaml with lockfileVersion: %f. Supported lockfile versions are %v", version, supportedVersions)
+}
+
+// DependencyMeta metadata for dependencies
+type DependencyMeta struct {
+	Injected bool   `yaml:"injected,omitempty"`
+	Node     string `yaml:"node,omitempty"`
+	Patch    string `yaml:"string,omitempty"`
+}
+
+// DecodePnpmLockfile parse a pnpm lockfile
+func DecodePnpmLockfile(contents []byte) (*PnpmLockfile, error) {
+	var lockfile PnpmLockfile
+	if err := yaml.Unmarshal(contents, &lockfile); err != nil {
+		return nil, errors.Wrap(err, "could not unmarshal lockfile: ")
+	}
+
+	if err := isSupportedVersion(lockfile.Version); err != nil {
+		return nil, err
+	}
+
+	return &lockfile, nil
+}
+
+// ResolvePackage Given a package and version returns the key, resolved version, and if it was found
+func (p *PnpmLockfile) ResolvePackage(name string, version string) (string, string, bool) {
+	resolvedVersion, ok := p.resolveSpecifier(name, version)
+	if !ok {
+		return "", "", false
+	}
+	key := fmt.Sprintf("/%s/%s", name, resolvedVersion)
+	if entry, ok := (p.Packages)[key]; ok {
+		var version string
+		if entry.Version != "" {
+			version = entry.Version
+		} else {
+			version = resolvedVersion
+		}
+		return key, version, true
+	}
+
+	return "", "", false
+}
+
+// AllDependencies Given a lockfile key return all (dev/optional/peer) dependencies of that package
+func (p *PnpmLockfile) AllDependencies(key string) (map[string]string, bool) {
+	deps := map[string]string{}
+	entry, ok := (p.Packages)[key]
+	if !ok {
+		return deps, false
+	}
+
+	for name, version := range entry.Dependencies {
+		deps[name] = version
+	}
+
+	for name, version := range entry.OptionalDependencies {
+		deps[name] = version
+	}
+
+	for name, version := range entry.PeerDependencies {
+		deps[name] = version
+	}
```

Comment:
Can these collide? Do we care if they do?

Reply:
From reading through pnpm I don't think these can collide. If a dep is marked optional then it won't appear in `dependencies`, peer deps are handled in the same manner. In the case that there's an optional peer it appears under peer with the `optional` attribute set true in `peerDependenciesMeta`

Reply:
Thanks for the explanation!